src/erasure.rs

```diff
@@ -161,263 +168,419 @@ pub fn decode_blocks(data: &mut [&mut [u8]], coding: &[&[u8]], erasures: &[i32])
-                let n = i % window.len();
-                if window[n].is_none() {
-                    trace!("coding block is null @ {}", n);
-                    return Ok(());
-                }
-                let w_l = window[n].clone().unwrap();
-                w_l.write().unwrap().set_size(max_data_size);
-                if w_l.write().unwrap().set_coding().is_err() {
-                    return Err(ErasureError::EncodeError);
-                }
-                coding_blobs.push(
-                    window[n]
-                        .clone()
-                        .expect("'coding_blobs' arr in pub fn generate_coding"),
-                );
+        let mut coding_blobs = Vec::with_capacity(NUM_CODING);
+
+        let coding_start = block_end - NUM_CODING;
+
+        for i in coding_start..block_end {
+            let n = i % window.len();
+            assert!(window[n].coding.is_none());
+
+            window[n].coding = Some(recycler.allocate());
+
+            let coding = window[n].coding.clone().unwrap();
+            let mut coding_wl = coding.write().unwrap();
+            for i in 0..max_data_size {
+                coding_wl.data[i] = 0;
             }
-            for b in &coding_blobs {
-                coding_locks.push(
-                    b.write()
-                        .expect("'coding_locks' arr in pub fn generate_coding"),
+            // copy index and id from the data blob
+            if let Some(data) = &window[n].data {
+                let data_rl = data.read().unwrap();
+
+                let index = data_rl.get_index().unwrap();
+                let id = data_rl.get_id().unwrap();
+
+                trace!(
+                    "{:x} copying index {} id {:?} from data to coding",
+                    debug_id,
+                    index,
+                    id
                 );
+                coding_wl.set_index(index).unwrap();
+                coding_wl.set_id(id).unwrap();
             }
-            for (i, l) in coding_locks.iter_mut().enumerate() {
-                trace!("i: {} coding: {} size: {}", i, l.data[0], max_data_size);
-                coding_ptrs.push(&mut l.data_mut()[..max_data_size]);
+            coding_wl.set_size(max_data_size);
+            if coding_wl.set_coding().is_err() {
+                return Err(ErasureError::EncodeError);
             }
 
-            generate_coding_blocks(coding_ptrs.as_mut_slice(), &data_ptrs)?;
-            debug!(
-                "consumed: {} data: {}:{} coding: {}:{}",
-                consumed,
-                block_start,
-                block_start + NUM_DATA,
-                coding_start,
-                coding_end
-            );
-            block_start += NUM_CODED;
+            coding_blobs.push(coding.clone());
         }
+
+        let data_locks: Vec<_> = data_blobs
+            .iter()
+            .map(|b| b.read().expect("'data_locks' of data_blobs"))
+            .collect();
```

Comment:
Drop the first 3 collects() to save 3 maallocs?

Reply:
I'm making 2 vectors, can only drop 2 collect()s...

Reply:
also: I think I have to hold the locks, so can't combine the map() calls

Reply:
Yeah, you'll need to keep that first collect() if all locks must be held before the next step.

Reply:
I was unsuccessful at merging the map() calls...  I'm gonna leave it unless you want me to keep pressing...